wallet-new/src/Cardano/Wallet/Kernel/Accounts.hs

```diff
@@ -0,0 +1,208 @@
+              -> AccountName
+              -- ^ The name for this account.
+              -> WalletId
+              -- ^ An abstract notion of a 'Wallet identifier
+              -> PassiveWallet
+              -> IO (Either CreateAccountError HdAccount)
+createAccount spendingPassword accountName walletId pw = do
+    let keystore = pw ^. walletKeystore
+    case walletId of
+         (WalletIdHdRnd hdRootId) -> do
+             mbEsk <- Keystore.lookup (WalletIdHdRnd hdRootId) keystore
+             case mbEsk of
+                  Nothing  -> return (Left $ CreateAccountKeystoreNotFound walletId)
+                  Just esk ->
+                      createHdRndAccount spendingPassword
+                                         accountName
+                                         esk
+                                         hdRootId
+                                         pw
+
+-- | Creates a new 'Account' using the random HD derivation under the hood.
+-- This code follows the same pattern of 'createHdRndAddress', but the two
+-- functions are "similarly different" enough to not make convenient generalise
+-- the code.
+createHdRndAccount :: PassPhrase
+                   -> AccountName
+                   -> EncryptedSecretKey
+                   -> HdRootId
+                   -> PassiveWallet
+                   -> IO (Either CreateAccountError HdAccount)
+createHdRndAccount _spendingPassword accountName _esk rootId pw = do
+    gen <- createSystemRandom
+    go gen 0
+    where
+        go :: GenIO -> Word32 -> IO (Either CreateAccountError HdAccount)
+        go gen collisions =
+            case collisions >= maxAllowedCollisions of
+                 True  -> return $ Left (CreateAccountHdRndAccountSpaceSaturated rootId)
+                 False -> tryGenerateAccount gen collisions
+
+        tryGenerateAccount :: GenIO
+                           -> Word32
+                           -- ^ The current number of collisions
+                           -> IO (Either CreateAccountError HdAccount)
+        tryGenerateAccount gen collisions = do
+            newIndex <- deriveIndex (flip uniformR gen) HdAccountIx HardDerivation
+            let hdAccountId = HdAccountId rootId newIndex
+                newAccount  = initHdAccount hdAccountId genesisCheckpoint &
+                              hdAccountName .~ accountName
+                db = pw ^. wallets
+            res <- update db (CreateHdAccount newAccount)
+            case res of
+                 (Left (CreateHdAccountExists _)) ->
+                     go gen (succ collisions)
+                 (Left (CreateHdAccountUnknownRoot _)) ->
+                     return (Left $ CreateAccountUnknownHdRoot rootId)
+                 Right () -> return (Right newAccount)
+
+        -- The maximum number of allowed collisions.
+        maxAllowedCollisions :: Word32
+        maxAllowedCollisions = 42
+
+        genesisCheckpoint :: Checkpoint
+        genesisCheckpoint = Checkpoint {
```

Comment:
Genesis is a misnomer here. It's not the genesis block. It's the birth of the new account.

Reply:
`birthdayCheckpoint` is an exquisite name I think 😛 I think I will use it.